### PR TITLE
Adds the stageprofi board for cob_light

### PR DESCRIPTION
**Pull request by thiagodefreitas**
- Some spaces/identation formatted due to issues with the old version

@ipa-nhg , @ipa-fmw Please test with the robot pc

### CODE REVIEW

**Comment by benmaidel** [comment]
@thiagodefreitas please remove all your tab to whitespace changes.

**Comment by thiagodefreitas** [comment]
@ipa-bnm My editor does it automatically, as it was configured for other project using the [ROS C++ Style guidelines](http://wiki.ros.org/CppStyleGuide) and [roslint recommendations](http://wiki.ros.org/roslint), where no tabs should be used, and indentation should be taken as 2 spaces.

Does that cause problems to you,somehow? If so, I can revert it. But otherwise, I would suggest that we keep the changes.

**Comment by benmaidel** [comment]
@thiagodefreitas No, this does not cause any problems on my side. The formating is ok. But the pull request is confusing because you can not see the actual implementation changes. It would be great if you could seperate the codeing style changes from the actual implementation changes into different pull request.

**Comment by mgruhler** [comment]
@ipa-bnm one of the great hidden features of github to avoid this issue I stumbled upon a while ago: add ?w=1 to the end of the URL to get rid of whitespace-only changes in the diffview

**Comment by benmaidel** [comment]
@ipa-mig hey, this is really nice :+1: But it is not possible to write source code comments in this view. I still think that seperating implementation from coding style is the right way to do it.

File: cob_light/msg/LightMode.msg [hunk 1]
```diff
@@ -14,4 +14,5 @@ float32 timeout 		# in s, requested mode will be executed for max timout s.
 int32 pulses 			# spezifies the amount of pulses which will be executed.
 			 			# eg: mode = flash, pulses = 2. Meens the light will flash two times
 int8 priority 			# priority [-20,20] default = 0. Modes with same or higher priorities will 
-						# be executed.
\ No newline at end of file
+						# be executed.
+int32 led_number
```
Comment (benmaidel) [inline cob_light/msg/LightMode.msg]: i think that an array of color values will be more effective than setting each led separately. perhaps something like: std_msgs/ColorRGBA[] colors

File: cob_light/ros/src/cob_light.cpp [hunk 2]
```diff
@@ -108,308 +109,330 @@ void shutdownCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result)
...
+      res.active_priority = p_modeExecutor->getExecutingPriority();
+      ROS_ERROR("Unsupported Color format. rgba values range is between 0.0 - 1.0");
+    }
+    else if(req.mode.mode == cob_light::LightMode::NONE)
+    {
+      p_modeExecutor->stop();
+      _color.a = 0;
+      if(_deviceDriver == "stageprofi")
+      {
+        std::cout << "Requested r" << req.mode.color.r;
```
Comment (benmaidel) [inline cob_light/ros/src/cob_light.cpp]: unnecessary debugging output

File: cob_light/ros/src/cob_light.cpp [hunk 3]
```diff
@@ -108,308 +109,330 @@ void shutdownCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result)
...
+    //ROS_DEBUG("Service Callback [Mode: %i with prio: %i freq: %f timeout: %f pulses: %i ]",
+    //	req.mode.mode, req.mode.priority, req.mode.frequency, req.mode.timeout, req.mode.pulses);
+
+    if(req.mode.color.r > 1.0 || req.mode.color.g > 1.0 || req.mode.color.b > 1.0 || req.mode.color.a > 1.0)
+    {
+      res.active_mode = p_modeExecutor->getExecutingMode();
+      res.active_priority = p_modeExecutor->getExecutingPriority();
+      ROS_ERROR("Unsupported Color format. rgba values range is between 0.0 - 1.0");
+    }
+    else if(req.mode.mode == cob_light::LightMode::NONE)
```
Comment (benmaidel) [inline cob_light/ros/src/cob_light.cpp:289]: LightMode NONE => all Lights off

File: cob_light/ros/src/cob_light.cpp [hunk 4]
```diff
@@ -108,308 +109,330 @@ void shutdownCallback(XmlRpc::XmlRpcValue& params, XmlRpc::XmlRpcValue& result)
...
+      p_modeExecutor->stop();
+      delete p_modeExecutor;
+    }
+    if(p_colorO != NULL)
+    {
+      delete p_colorO;
+    }
+  }
+
+  void topicCallback(std_msgs::ColorRGBA color)
```
Comment (benmaidel) [inline cob_light/ros/src/cob_light.cpp:258]: i think stageprofi should also be handelt in this case?

File: cob_light/msg/LightMode.msg (hunk 1 above)
Comment (thiagodefreitas) [inline cob_light/msg/LightMode.msg]: The possibility to control each LED separately is required for our use cases.

File: cob_light/ros/src/cob_light.cpp (hunk 4 above)
Comment (thiagodefreitas) [inline cob_light/ros/src/cob_light.cpp:258]: I may have misunderstood, but I think the p_color0 and the p_modeExecutor can be destroyed for the stageprofi there as well, without any additional if.

File: cob_light/ros/src/cob_light.cpp (hunk 3 above)
Comment (thiagodefreitas) [inline cob_light/ros/src/cob_light.cpp:289]: Fixed

File: cob_light/ros/src/cob_light.cpp (hunk 4 above)
Comment (benmaidel) [inline cob_light/ros/src/cob_light.cpp:258]: Yes, but i meant the topicCallback function. I think in this function the stageprofi needs to be handelt?

File: cob_light/msg/LightMode.msg (hunk 1 above)
Comment (benmaidel) [inline cob_light/msg/LightMode.msg]: Then we should probably come up with a better solution. Calling x times the same action/service only to set all lights is horrible. Imagine if we have 512 leds or even more...
What about a index array and a value array?
uint32 [] led_numbers
std_msgs/ColorRGBA[] colors

File: cob_light/msg/LightMode.msg [hunk 5]
```diff
@@ -14,4 +14,6 @@ float32 timeout 		# in s, requested mode will be executed for max timout s.
 int32 pulses 			# spezifies the amount of pulses which will be executed.
 			 			# eg: mode = flash, pulses = 2. Meens the light will flash two times
 int8 priority 			# priority [-20,20] default = 0. Modes with same or higher priorities will 
-						# be executed.
\ No newline at end of file
+						# be executed.
+uint32[] led_numbers
```
Comment (floweisshardt) [inline cob_light/msg/LightMode.msg:18]: @thiagodefreitas: please add some comment here to indicate for what led_numbers and colors can be used. (e.g. which mode, default value, ...). Should be similar to the comments for the other message entries above.

**Comment by NadiaHG** [comment]
Visual Review: OK

**Comment by floweisshardt** [comment]
visual review: OK

**Pull request merged**